packages/astro/src/core/build/static-build.ts

```diff
@@ -231,85 +235,92 @@ async function ssrBuild(
+	const clientOutput = await builder.build(builder.environments.client);
+
+	return { ssrOutput, prerenderOutput, clientOutput };
 }
 
-async function clientBuild(
+type MutateChunk = (chunk: vite.Rollup.OutputChunk, targets: string[], newCode: string) => void;
+
+async function runManifestInjection(
 	opts: StaticBuildOptions,
 	internals: BuildInternals,
-	input: Set<string>,
-	container: AstroBuildPluginContainer,
+	ssrOutputs: vite.Rollup.RollupOutput[],
+	clientOutputs: vite.Rollup.RollupOutput[],
+	prerenderOutputs: vite.Rollup.RollupOutput[],
 ) {
-	const { settings, viteConfig } = opts;
-	const ssr = settings.buildOutput === 'server';
-	const out = ssr ? settings.config.build.client : getOutDirWithinCwd(settings.config.outDir);
-
-	// Nothing to do if there is no client-side JS.
-	if (!input.size) {
-		// If SSR, copy public over
-		if (ssr && fs.existsSync(settings.config.publicDir)) {
-			await fs.promises.cp(settings.config.publicDir, out, { recursive: true, force: true });
+	const mutations = new Map<
+		string,
+		{
+			targets: string[];
+			code: string;
 		}
-
-		return null;
-	}
-
-	const { lastVitePlugins, vitePlugins } = await container.runBeforeHook('client', input);
-	opts.logger.info('SKIP_FORMAT', `\n${bgGreen(black(' building client (vite) '))}`);
-
-	const viteBuildConfig: vite.InlineConfig = {
-		...viteConfig,
-		build: {
-			target: 'esnext',
-			...viteConfig.build,
-			emptyOutDir: false,
-			outDir: fileURLToPath(out),
-			copyPublicDir: ssr,
-			rollupOptions: {
-				...viteConfig.build?.rollupOptions,
-				input: Array.from(input),
-				output: {
-					format: 'esm',
-					entryFileNames: `${settings.config.build.assets}/[name].[hash].js`,
-					chunkFileNames: `${settings.config.build.assets}/[name].[hash].js`,
-					assetFileNames: `${settings.config.build.assets}/[name].[hash][extname]`,
-					...viteConfig.build?.rollupOptions?.output,
-				},
-				preserveEntrySignatures: 'exports-only',
-			},
-		},
-		plugins: [...vitePlugins, ...(viteConfig.plugins || []), ...lastVitePlugins],
-		envPrefix: viteConfig.envPrefix ?? 'PUBLIC_',
-		base: settings.config.base,
+	>();
+
+	const mutate: MutateChunk = (chunk, targets, newCode) => {
+		chunk.code = newCode;
+		mutations.set(chunk.fileName, {
+			targets,
+			code: newCode,
+		});
 	};
 
-	const updatedViteBuildConfig = await runHookBuildSetup({
-		config: settings.config,
-		pages: internals.pagesByKeys,
-		vite: viteBuildConfig,
-		target: 'client',
-		logger: opts.logger,
+	await manifestBuildPostHook(opts, internals, {
+		ssrOutputs,
+		clientOutputs,
+		prerenderOutputs,
+		mutate,
 	});
 
-	const buildResult = await vite.build(updatedViteBuildConfig);
-	return buildResult;
+	await writeMutatedChunks(opts, mutations, prerenderOutputs);
 }
 
-async function runPostBuildHooks(
-	container: AstroBuildPluginContainer,
-	ssrOutputs: vite.Rollup.RollupOutput[],
-	clientOutputs: vite.Rollup.RollupOutput[],
+async function writeMutatedChunks(
```

Comment:
The function's name is very cryptic. I think this is a good opportunity to write some docs and explain what we use this function for.